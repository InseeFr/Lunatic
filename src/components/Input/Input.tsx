import classnames from 'classnames';
import './Input.scss';
import type { LunaticComponentProps } from '../type';
import { Label } from '../shared/Label/Label';
import { slottableComponent } from '../shared/HOC/slottableComponent';
import {
	ComponentErrors,
	getComponentErrors,
} from '../shared/ComponentErrors/ComponentErrors';
import { Declarations } from '../shared/Declarations/Declarations';
import type { LunaticError } from '../../use-lunatic/type';

export function Input({
	handleChange,
	response,
	errors,
	...props
}: LunaticComponentProps<'Input'>) {
	return (
		<CustomInput
			{...props}
			onChange={(v) => handleChange(response, v)}
			errors={getComponentErrors(errors, props.id)}
		/>
	);
}

type CustomProps = Omit<
	LunaticComponentProps<'Input'>,
	'response' | 'handleChange' | 'errors'
> & {
	onChange: (v: string) => void;
	errors?: LunaticError[];
};

export const CustomInput = slottableComponent<CustomProps>('Input', (props) => {
	const {
		value,
		onChange,
		disabled,
		required,
		maxLength,
		label,
		description,
		id,
		errors,
		readOnly,
		declarations,
	} = props;
	const labelId = `label-${id}`;
	return (
		<div className={classnames('lunatic-input')}>
			<Label htmlFor={id} id={labelId} description={description}>
				{label}
			</Label>
			<Declarations
				type="AFTER_QUESTION_TEXT"
				declarations={declarations}
				id={id}
			/>
			<input
				id={id}
				aria-labelledby={labelId}
				autoComplete="off"
				type="text"
				disabled={disabled}
				readOnly={readOnly}
				value={(value ?? '').toString()}
				onChange={(e) => onChange(e.target.value)}
				aria-required={required}
				required={required}
				maxLength={maxLength}
				aria-invalid={!!errors}
			/>
			<ComponentErrors errors={errors} />
		</div>
	);
});
