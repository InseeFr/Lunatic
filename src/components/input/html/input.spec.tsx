import { fireEvent, render, screen } from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import Input from './input';
describe('Input', () => {
	const mockOnChange = vi.fn();

	beforeEach(() => {
		mockOnChange.mockClear();
	});

	it('renders without crashing', () => {
		const { container } = render(
			<Input
				value={'input'}
				id="input"
				onChange={mockOnChange}
				errors={undefined}
				description="description"
			/>
		);
		expect(container).toMatchSnapshot();
	});

	it('renders label and input', () => {
		const labelText = 'Enter a number';
		const { container } = render(
			<Input
				id="number"
				label={labelText}
				onChange={mockOnChange}
				errors={undefined}
			/>
		);

		const label = screen.getByText(labelText);
		const input = container.querySelector('input[type="text"]');

		expect(label).toBeInTheDocument();
		expect(input).toBeInTheDocument();
	});

	it('calls onChange with parsed value', () => {
		const { container } = render(
			<Input id="number" onChange={mockOnChange} errors={undefined} />
		);

		const input = container.querySelector('input[type="text"]')!;
		fireEvent.change(input, { target: { value: 'Dupont' } });

		expect(mockOnChange).toHaveBeenCalledTimes(1);
		expect(mockOnChange).toHaveBeenCalledWith('Dupont');
	});

	it('disables input when disabled prop is true', () => {
		const { container } = render(
			<Input id="number" disabled onChange={mockOnChange} errors={undefined} />
		);

		const input = container.querySelector('input[type="text"]');
		expect(input).toBeDisabled();
	});

	it('should handle readOnly', () => {
		const { container } = render(
			<Input
				id="number"
				value="toto"
				readOnly
				onChange={mockOnChange}
				errors={undefined}
			/>
		);
		expect(container).toMatchSnapshot();

		const input = container.querySelector('input[type="text"]');
		expect(input).toHaveAttribute('readonly');
		(input as HTMLElement).focus();
		expect(input).toHaveFocus();
		expect(input).toHaveValue('toto');
	});
});
