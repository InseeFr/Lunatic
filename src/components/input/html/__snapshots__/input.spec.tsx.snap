// Vitest Snapshot v1

exports[`Input > renders without crashing 1`] = `
<div>
  <div
    class="lunatic-input"
  >
    <input
      aria-labelledby="label-input"
      autocomplete="off"
      id="input"
      type="text"
      value="input"
    />
  </div>
</div>
`;

exports[`Input > should handle readOnly 1`] = `
<div>
  <div
    class="lunatic-input"
  >
    <input
      aria-labelledby="label-number"
      autocomplete="off"
      id="number"
      readonly=""
      type="text"
      value="toto"
    />
  </div>
</div>
`;
